### ISSUE TITLE
Differences between openshift mysql image and RHSCL image and other things to do

### ISSUE THREAD

**Issue by hhorak** [sclorg/rhscl2dockerfile#17]
This issue should track what are the differences between openshift mysql image available at https://github.com/openshift/mysql and RHSCL image generated here. The differences we care are from user's PoV, not differences in implementation.

So far I've identified these differences:
* general_log is on -- I believe this is not something all users should have enabled, since the queries may include confidential data like passwords and stuff
* master/slave replication -- we agreed on the meeting we don't want to ship it in general docker image, right? it should only serve as example of what is possible to do with the docker image, so what about to keep these scripts in some `contrib` directory, document how it can be used, but not `add` this functionality in the docker image?
* my.cnf generation from environment variables -- it is possible to pass the variables now even in RHSCL docker image, using `docker run -e MYSQL_CONFIG_key_buffer_size=8388608 ...` -- advantages of this way are that there is no docker image update needed for adding a new option, we don't have to care what is the default value of particular option and user knows precisely what option is being set.
* logging to errors to stderr -- this particularly means not to set --log_error option in the my.cnf

@mnagy @mfojtik @bparees Let's get the above issues solved. Are there any other differences from user PoV I forgot or didn't realize?

**Comment by hhorak** [comment]
https://github.com/sclorg/rhscl2dockerfile/issues/13 is connected to this issue, but the security impact on `general_log` enabled by default is still valid (and is specific for mysql docker image).

**Comment by bparees** [comment]
Thanks for writing this up, definitely a good idea to compare and get to a common design.

general_log: agree, we should not have that on by default but it should be configurable via an ENV variable.

master/slave: not sure why we wouldn't ship the scripts in the image so the image can be run in that mode, as long as we're reasonably comfortable we've done it "right"...  @mfojtik?

handling env variable config:  it's a nice idea but it won't work for all languages/frameworks (eg anything that uses an XML based config, or that can handle a config value being repeated in the file).  I think there were other arguments against this too but I don't remember them offhand... @mfojtik?  If what I listed are the only arguments against it, I guess we could do it for this image (but not others) but then I wonder if that's a good idea, if we can't be consistent about it.

Also if there are any config values we'd want to setup based on the container parameters (eg sizing a threadpool or cache based on cpu/memory allocations) we'd need more complex config management logic anyway.  We intend to have that in at least some of the images, not sure if mysql would be one of those.  (the idea being we'd want to default to a calculated value but still allow users to explicitly set the value instead).  That would be incorporated with this scheme, but i think it ends up being more complicated.

logging errors to stderr: seems reasonable to me that we'd want to log errros to stderr, though i don't know that it really matters since i believe both stdout+stderr end up in the docker log together and i'm not sure there's any indication which stream they originally came from.

**Comment by praiskup** [comment]
@bparees
> master/slave: not sure why we wouldn't ship the scripts in the image so the
> image can be run in that mode, as long as we're reasonably comfortable we've
> done it "right"... @mfojtik?

There are projects dedicated to master/slave replication.  I really doubt we
will be able to guarantee something about the "default" configuration.

Talking about PostgreSQL container, I have no problem to put scripts under
contrib/master-slave/ in git repository (or something like that) if you wish,
which you would be able to use like
```
-v contrib/master-slave/slave:/usr/share/cont-volume/postgresql/postinitdb
```

But the approach you propose is wrong.  You reimplement the configuration
already implemented!  This does not sound like something we will have
resources to maintain.

Or, ideally, if you guys are able to maintain such things, we could do it as
separate project injectable as additional layer?  That would be _so_ good.

> handling env variable config: it's a nice idea but it won't work for all
> languages/frameworks

If the general idea works for at least PostgreSQL or MariaDB, I really doubt
we should choose another approach.  That would be just huge wasting of
resources, why?

I don't really think its worth doing respin of dozens of images with PG just
because we decided to support yet another variable name for yet another
configuration option.

> (eg anything that uses an XML based config, or that can handle a config
> value being repeated in the file).

Then we should find a good generic solution too.  Mounting volumes with
configuration does not sound like a bad idea.

> I think there were other arguments against this too but I don't remember
> them offhand... @mfojtik? If what I listed are the only arguments against
> it, I guess we could do it for this image (but not others) but then I wonder
> if that's a good idea, if we can't be consistent about it.

We need to do as less work as possible, even if we choose
non-consistent approach among our images.

**Comment by mfojtik** [comment]
On May 19, 2015 9:51 PM, "praiskup" <notifications@github.com> wrote:
>
> @bparees
>>
>> master/slave: not sure why we wouldn't ship the scripts in the image so
the
>> image can be run in that mode, as long as we're reasonably comfortable
we've
>> done it "right"... @mfojtik?
>
> There are projects dedicated to master/slave replication. I really doubt
we
> will be able to guarantee something about the "default" configuration.

We should at least try to provide an example or base our users can start
from. Kubernetes is complex tool and having a simply to understand example
in place will be a huge help.

> Talking about PostgreSQL container, I have no problem to put scripts under
> contrib/master-slave/ in git repository (or something like that) if you
wish,
> which you would be able to use like
>
> -v
contrib/master-slave/slave:/usr/share/cont-volume/postgresql/postinitdb

Mounting directories with config files is no go for Kubernetes, as we
already discussed that (many times)

>
> But the approach you propose is wrong. You reimplement the configuration
> already implemented! This does not sound like something we will have
> resources to maintain.

What reimplementation are you talking about? And what approach you are
referencing to? The one I am aware of was done by one of our partners....
Which is another reason why we should care.

>
> Or, ideally, if you guys are able to maintain such things, we could do it
as
> separate project injectable as additional layer? That would be so good.
>>
>> handling env variable config: it's a nice idea but it won't work for all
>> languages/frameworks
>
> If the general idea works for at least PostgreSQL or MariaDB, I really
doubt
> we should choose another approach. That would be just huge wasting of
> resources, why?
>
> I don't really think its worth doing respin of dozens of images with PG
just
> because we decided to support yet another variable name for yet another
> configuration option.

We are not building images that allow total customization or be 100%
generic. That is misunderstanding of how Docker works.

>>
>> (eg anything that uses an XML based config, or that can handle a config
>> value being repeated in the file).
>
> Then we should find a good generic solution too. Mounting volumes with
> configuration does not sound like a bad idea.

It is very bad idea.

>>
>> I think there were other arguments against this too but I don't remember
>> them offhand... @mfojtik? If what I listed are

Yes. What will happen if in new version, a config setting name is changes
or removed? How your solution will help people upgrade?

the only arguments against
>> it, I guess we could do it for this image (but not others) but then I
wonder
>> if that's a good idea, if we can't be consistent about it.
>
> We need to do as less work as possible, even if we choose
> non-consistent approach among our images.

Disagree. Env vars is simplest possible approach we can use generically for
all images.

**Comment by bparees** [comment]
> I don't really think its worth doing respin of dozens of images with PG just
> because we decided to support yet another variable name for yet another
> configuration option.

why will we have "dozens" of PG images?


> Then we should find a good generic solution too. Mounting volumes with
> configuration does not sound like a bad idea.

that sounds like a very bad idea.  it makes it incredibly difficult to ensure consistent configuration across deployments, it requires more management to setup those volumes and ensure they are mounted, etc.  Config should absolutely be injected in a way that can be defined in scripts or in the case of openshift, defined in a template or other config json.

> We need to do as less work as possible, even if we choose
> non-consistent approach among our images.

inconsistency in implementation tends to create more work, not less.  you end up with different bugs in different implementations and more work for people trying to come up to speed on different images that use different mechanisms.  And frankly the existing mechanism used in the openshift images is not that much work.

**Comment by mfojtik** [comment]
On May 19, 2015 10:12 PM, "Ben Parees" <notifications@github.com> wrote:
>>
>> I don't really think its worth doing respin of dozens of images with PG
just
>> because we decided to support yet another variable name for yet another
>> configuration option.
>
> why will we have "dozens" of PG images?

2 versions of PG multiply by  4 (rhel6,rhel7,fedora,centos) is 8 images....
also we are building tools that automatize building them, right?

>>
>> Then we should find a good generic solution too. Mounting volumes with
>> configuration does not sound like a bad idea.
>
> that sounds like a very bad idea. it makes it incredibly difficult to
ensure consistent configuration across deployments, it requires more
management to setup those volumes and ensure they are mounted, etc. Config
should absolutely be injected in a way that can be defined in scripts or in
the case of openshift, defined in a template or other config json.
>>
>> We need to do as less work as possible, even if we choose
>> non-consistent approach among our images.
>
> inconsistency in implementation tends to create more work, not less. you
end up with different bugs in different implementations and more work for
people trying to come up to speed on different images that use different
mechanisms. And frankly the existing mechanism used in the openshift images
is not that much work.

**Comment by praiskup** [comment]
@mfojtik wrote:
> Mounting directories with config files is no go for Kubernetes, as we already
> discussed that (many times)

Yes, but nobody said what is wrong with HostDir kubernetes configuration
so far.  And we discussed it privately, so its worth public discussion.

> What reimplementation are you talking about?

Reimplementation: You need to check that variable dedicated to some
configuration option meets the requirement (regexp, max value, etc.).  Why
we should reimplement that if the installed tool (PostgreSQL e.g.)
implements that checks already?

> And what approach you are referencing to?

One supported env variable for one configuration option.  That approach
does not scale.

> We are not building images that allow total customization or be 100%
> generic. That is misunderstanding of how Docker works.

If I give user a generic way to configure image, I do it wrong in Docker?
Citation needed.

> Yes. What will happen if in new version, a config setting name is
> changes or removed? How your solution will help people upgrade?

I don't understand this argument.  If something changes - you just don't
have to remove some variable you invented before.

> > We need to do as less work as possible, even if we choose
> > non-consistent approach among our images.
>
> Disagree. Env vars is simplest possible approach we can use generically
> for all images.

Not for all cases.  If there does not exist configuration manager e.g. for
httpd yet (in real world), why we should pretend we can handle stuff like
that in our containers?  And if there is some, shouldn't we use that?

Pavel

**Comment by mfojtik** [comment]
On May 19, 2015 10:25 PM, "praiskup" <notifications@github.com> wrote:
>
> @mfojtik wrote:
>>
>> Mounting directories with config files is no go for Kubernetes, as we
already
>> discussed that (many times)
>
> Yes, but nobody said what is wrong with HostDir kubernetes configuration
> so far. And we discussed it privately, so its worth public discuss

How you will use that?

>>
>> What reimplementation are you talking about?
>
> Reimplementation: You need to check that variable dedicated to some
> configuration option meets the requirement (regexp, max value, etc.). Why
> we should reimplement that if the installed tool (PostgreSQL e.g.)
> implements that checks already?

Just to give users hint what they did wrong or what is required to be set
for the image to operate successfully and be useful?

If you think some of the validations we have are redundant then we can
remove them.

>>
>> And what approach you are referencing to?
>
> One supported env variable for one configuration option. That approach
> does not scale.

You keep repeating that. What you actually mean with "scale"? I think it
scales well and it might not map to a setting in every case... It also
might help upgrading.

>>
>> We are not building images that allow total customization or be 100%
>> generic. That is misunderstanding of how Docker works.
>
> If I give user a generic way to configure image, I do it wrong in Docker?
> Citation needed.

It depends on audience. Our users or Red Hat users don't want to build
stuff them self if we can provide some added value already?
If you want to play with Docker, then you might start with one of the
generic images on Hub, but in the end you will not be using that.

>>
>> Yes. What will happen if in new version, a config setting name is
>> changes or removed? How your solution will help people upgrade?
>
> I don't understand this argument. If something changes - you just don't
> have to remove some variable you invented before.
>>>
>>> We need to do as less work as possible, even if we choose
>>> non-consistent approach among our images.
>>
>> Disagree. Env vars is simplest possible approach we can use generically
>> for all images.
>
> Not for all cases. If there does not exist configuration manager e.g. for
> httpd yet (in real world), why we should pretend we can handle stuff like
> that in our containers? And if there is some, shouldn't we use that?

If you invent a configuration management tool for containers running in
cluster, will you maintain it?

**Comment by bparees** [comment]
> Yes, but nobody said what is wrong with HostDir kubernetes configuration
> so far. And we discussed it privately, so its worth public discussion.

hostdir is very bad since it means your pod cannot migrate from one node to another (the hostdir won't be there) and you would still need an external mechanism to go create the hostdir with the config in it.  it's definitely not the right way to use kubernetes.  Hostdir is basically only valid as a scratch volume that starts empty and is used to write throwaway data.

> Reimplementation: You need to check that variable dedicated to some
> configuration option meets the requirement (regexp, max value, etc.). Why
> we should reimplement that if the installed tool (PostgreSQL e.g.)
> implements that checks already?

there's no requirement to do such checks... you can pass the config value directly to the framework and let it report errors, as you describe, with either mechanism.

> I don't understand this argument. If something changes - you just don't
> have to remove some variable you invented before.

if the name of the config property changes, and you (As an image author) want to seemlessly migrate users from the old version to the new version, you need to do what we are doing.  Otherwise you break users who are referencing the old config property name.

**Comment by praiskup** [comment]
> > why will we have "dozens" of PG images?

> 2 versions of PG multiply by  4 (rhel6,rhel7,fedora,centos) is 8 images....

You forget about multiple versions of RHSCL and that we have more versions of
fedora or centos.  And you can think of additional layers, of course.

> > Then we should find a good generic solution too. Mounting volumes with
> > configuration does not sound like a bad idea.
>
> that sounds like a very bad idea. it makes it incredibly difficult to ensure
> consistent configuration across deployments,

Why?

> it requires more management to setup those volumes and ensure they are
> mounted, etc.

I don't think this is something we should solve *in* images.  But rather
by the tooling around.

> Config should absolutely be injected in a way that can be defined in
> scripts or in the case of openshift, defined in a template or other
> config json.

This is not scalable.  You need to iterate forever.  Users are blocked to
use configuration options they want to use, *now*.

> > We need to do as less work as possible, even if we choose
> > non-consistent approach among our images.
>
> inconsistency in implementation tends to create more work, not less.

Come on.  If something works in an absolute generic way for some
container, why we should add redundant code there?  Thats way too much
work.

> And frankly the existing mechanism used in the openshift images is not
> that much work.

Its easy to implement the variable and then make somebody else to maintain
that forever.  And frankly, it will be way too much work, without any
boundaries, without a way back.

Pavel

**Comment by praiskup** [comment]
> > Yes, but nobody said what is wrong with HostDir kubernetes configuration
> > so far. And we discussed it privately, so its worth public discussion.
>
> hostdir is very bad since it means your pod cannot migrate from one node to
> another (the hostdir won't be there) and you would still need an external
> mechanism to go create the hostdir with the config in it.

No!  Why you can't have compatible nodes?  That will have installed package
having proper configuration?  The same way as it has installed kubernetes?

> the right way to use kubernetes. Hostdir is basically only valid as a scratch
> volume that starts empty and is used to write throwaway data.

Or statically prepackaged data, which are also read-only and throwaway.

>> Reimplementation: You need to check that variable dedicated to some
>> configuration option meets the requirement (regexp, max value, etc.). Why
>> we should reimplement that if the installed tool (PostgreSQL e.g.)
>> implements that checks already?
>
>there's no requirement to do such checks... you can pass the config value
>directly to the framework and let it report errors, as you describe, with
>either mechanism.

Then, why we should have unlimited number of variables if we can, at least in
some cases, have only one variable?

> >  I don't understand this argument. If something changes - you just don't
> >  have to remove some variable you invented before.
>
> if the name of the config property changes, and you (As an image author) want
> to seemlessly migrate users from the old version to the new version, you need
> to do what we are doing.

This is science fiction.  We can have migration tools, yes.  But it is
separate topic - guys from preupgrade-assistant could comment.

But we definitely can't keep API if upstream decides to remove something
between major versions or so.  Well, when upstream decides to change API, it
must definitely have a good reason which should match also container scenario.

Pavel

**Comment by mfojtik** [comment]
On May 19, 2015 10:44 PM, "praiskup" <notifications@github.com> wrote:
>>>
>>> why will we have "dozens" of PG images?
>>
>> 2 versions of PG multiply by 4 (rhel6,rhel7,fedora,centos) is 8
images....
>
> You forget about multiple versions of RHSCL and that we have more
versions of
> fedora or centos. And you can think of additional layers, of course

Isn't this why we are building automation around this?

>>>
>>> Then we should find a good generic solution too. Mounting volumes with
>>> configuration does not sound like a bad idea.
>>
>> that sounds like a very bad idea. it makes it incredibly difficult to
ensure
>> consistent configuration across deployments,
>
> Why?
>>
>> it requires more management to setup those volumes and ensure they are
>> mounted, etc.
>
> I don't think this is something we should solve in images. But rather
> by the tooling around.

Who is going to write that tooling?

>>
>> Config should absolutely be injected in a way that can be defined in
>> scripts or in the case of openshift, defined in a template or other
>> config json.
>
> This is not scalable. You need to iterate forever. Users are blocked to
> use configuration options they want to use, now.

What is scalable? Users have way to use whatever they want, nothing
prevents you from adding layer that overrides default config. But that is
not the experience OpenShift or some Red Hat customers/users wants.

>>>
>>> We need to do as less work as possible, even if we choose
>>> non-consistent approach among our images.
>>
>> inconsistency in implementation tends to create more work, not less.
>
> Come on. If something works in an absolute generic way for some
> container, why we should add redundant code there? Thats way too much
> work.

If something works in absolute generic way why people use Docker images?
And how are they useful? If you have to write all configuration and
bootstrapping yourself?

>>
>> And frankly the existing mechanism used in the openshift images is not
>> that much work.
>
> Its easy to implement the variable and then make somebody else to maintain
> that forever. And frankly, it will be way too much work, without any
> boundaries, without a way back.

There is no forever and I don't see why we should have way back and what is
the problem here. Users can safely ignore them and use images in a way you
want already.

**Comment by praiskup** [comment]
@mfojtik wrote:
> Who is going to write that tooling?

I was about to think that you already do that in OpenShift.  If you mount data volume for
database, you need to ensure the same stuff you would need to ensure for mounted
configuration.

Who is going to add and maintain the (with time) unlimited number of variables for each
particular configuration option of XXX component?

> If something works in absolute generic way why people use Docker images?

I did not speak that generic .. :( I don't feel this is fair point.

> There is no forever and I don't see why we should have way back and what is
> the problem here.

There actually is "forever".  And I'm not the one who mentioned that we could
keep the variable even though upstream decided to remove the relevant
configuration option.

--
TBH, I'm getting lost in this thread and we need to speak concretely.

Could anybody interested in look at
https://github.com/fedora-cloud/Fedora-Dockerfiles/pull/105
whether that is the way we should *never go* or find possible problems?

To make it clear:  There is one variable "POSTGRESQL_CONFIG", that takes
`key = value [; key = value ]`.  Is there a problem with that approach?
Note that this approach was invented just because you required configuration
via environment variables, not mounting.  I would still prefer mounting.

If the `POSTGRESQL_MAX_CONNECTIONS` and `POSTGRESQL_SHARED_BUFFER` variables
(and all the future variables) are absolute must for OpenShift, would that be
a problem to implement (and maintain) those variables as separate shell script?
That script could be added into /usr/share/cont-layer/postgresql/postinitdb in
openshift's additional layer.

**Comment by hhorak** [comment]
Since mounting the config is no-go for openshift, let's focus on "env. variables" way to change configuration then.
Also we seem to agree that usability is important from users' PoV. That said, the generic way @praiskup seems better to me, since users will shortly understand the concept and will be able to do much more; comparing to the "fixed variables" approach where users would have to first find out, which name of the variable to use and then they will probably miss some other variables.

**Comment by mfojtik** [comment]
@hhorak as @bparees mentioned earlier, the 'generic way' that @praiskup suggested is not that generic as it doesn't work for XML configurations and JSON configuration files, right? Also it won't work in cases when the setting will be something like:

`MYSQL_CONTAINER_SIZE=medium`

Which can be used to calculate the optimal configuration settings for MySQL/PostgreSQL/... based on the the size of the container. That will affect multiple fields... (and that is just an example)...

**Comment by praiskup** [comment]
@mfojtik wrote:
> @hhorak as @bparees mentioned earlier, the 'generic way' that @praiskup
> suggested is not that generic as it doesn't work for XML configurations
> and JSON configuration files, right?

Well, it works for PostgreSQL where we don't care about xml & json, right?

> Also it won't work in cases when the setting will be something like:
> MYSQL_CONTAINER_SIZE=medium
> Which can be used to calculate the optimal configuration settings for
> MySQL/PostgreSQL/... based on the the size of the container. That will affect
> multiple fields... (and that is just an example)...

This sounds like orthogonal problem, however.  The RFE for
MYSQL_CONTAINER_SIZE would deal with "defaults".

**Comment by bparees** [comment]
it also won't work for any framework that doesn't tolerate duplicate entries of config values since we'd want to specify a default value in the config file we supply, and then the user would set the ENV variable which would result in a second entry for that property being appended to the config file.

Some frameworks use the first value, some use the second, and probably some will even throw a validation error.

So depending on the framework and whether the config file provided a default value entry you'd need to either:
1) prefix the values
2) append the values
3) replace the values

**Comment by mfojtik** [comment]
@bparees @hhorak also users don't want to modify **all** configuration settings.

80% of users (based on the years of OpenShift v2 experience) are happy with being able to customize the fields that are required for their applications (locale settings, TZ, datestyle, etc...) and they are fine with the default configuration or configuration that the provider created for them with optimal settings...

 There is literally **no need** to invent a complex or 'generic' way to configure things as nobody is asking for that. As I said before, the 20% of cases, when you require something special are covered already as you can just add layer with your configuration and ignore all ENV variables all together.

**Comment by praiskup** [comment]
@praiskup wrote:
> TBH, I'm getting lost in this thread and we need to speak concretely.
>
> Could anybody interested in look at
> https://github.com/fedora-cloud/Fedora-Dockerfiles/pull/105
> whether that is the way we should *never go* or find possible problems?
>
> To make it clear:  There is one variable "POSTGRESQL_CONFIG", that takes
> `key = value [; key = value ]`.  Is there a problem with that approach?
> Note that this approach was invented just because you required configuration
> via environment variables, not mounting.  I would still prefer mounting.
>
> If the `POSTGRESQL_MAX_CONNECTIONS` and `POSTGRESQL_SHARED_BUFFER` variables
> (and all the future variables) are absolute must for OpenShift, would that be
> a problem to implement (and maintain) those variables as separate shell script?
> That script could be added into /usr/share/cont-layer/postgresql/postinitdb in
> openshift's additional layer.

Could we solve that ^^^ points?  Thanks.

**Comment by mfojtik** [comment]
@praiskup `POSTGRESQL_CONFIG` works for postgresql right? In MySQL, the config file has sections, so I wonder how you will solve that (section.key=value?). Then you have Tomcat, which is XML config file (pretty ugly one). You your approach is generic to that?

We already tried to do something similar for OpenShift v2 PostreSQL cartridge. You know what happened? People use it and they made mistakes (typos, wrong values, etc) and we give them errors that they will then fill in BZ. 
Your approach will basically follow the same pattern (long term). People will use it to configure PostgreSQL server. When they use incorrect syntax or provide wrong value (that might be wrong because it does not fit the container size), they get errors and then they will go to BZ...

@praiskup another thing, I mentioned earlier is, how this would help people upgrade? Let say there is a new minor version of PostgreSQL that deprecate some value in postgresql.conf. If you upgrade, postgresql containers wont start and you as image provider will go tell 10000 users to update the value...

@praiskup the main goal of PaaS service is to protect users from this to happen. **We** as provider care about your application and we will try to prevent breaking you when we upgrade our images (instead of failing to start your container). Having simple POSTRESQL_FOO=value also means that YOU can validate the value and fix it (or default) if it is no longer valid or it was renamed.

**Comment by praiskup** [comment]
@mfojtik wrote:
> @praiskup POSTGRESQL_CONFIG works for postgresql right?

Thats everything I try to solve now.  Thanks.

> In MySQL, the config file has sections, so I wonder how you will solve that
> (section.key=value?).

There should not be a problem with this, probably.

> Then you have Tomcat, which is XML config file (pretty ugly one). How is your
> approach generic to that?

It does not try to be that generic.

> We already tried to do something similar for OpenShift v2 PostreSQL cartridge.
> You know what happened? People use it and they made mistakes (typos, wrong
> values, etc) and we give them errors that they will then fill in BZ.

That is not a problem to me.  Customers will do typos even while spelling
variable names, without warning (non-supported env variable does not have an
effect).

> Your approach will basically follow the same pattern (long term). People will
> use it to configure PostgreSQL server. When they use incorrect syntax or
> provide wrong value (that might be wrong because it does not fit the container
> size), they get errors and then they will go to BZ...

Well, I don't see the problem with bugzilla here.  I can react and say its a bug
and fix that, or say it is NOTABUG.

> @praiskup another thing, I mentioned earlier is, how this would help people
> upgrade?

Upgrade of PostgreSQL major versions is separate topic.

> Let say there is a new minor version of PostgreSQL that deprecate some value
> in postgresql.conf.

Has this ever happened (citation needed).  Don't predict that.  PostgreSQL is
fine upstream to change _that way_ API for minor versions.  This needs to be
solved somewhere else then in container's logic (e.g. upstream, or for system
package - because there will be the same inconsistency).

> If you upgrade, postgresql containers wont start and you as image provider
> will go tell 10000 users to update the value manually?

Using your argument;  you are trying to solve hypothetical problem.  We can't
update component and break API.  Thats not the way RHEL works ;).

To the three "parts" of answer above - the configuration variables are used
only during initdb time..  which will not happen in migrations.  Ugrades to
new major version is topic to be solved in future (and you argumented some
time ago that we definitely do not want it).

> @praiskup the main goal of PaaS service is to protect users from this to
> happen. We as provider care about your application and we will try to
> prevent breaking you when we upgrade our images (instead of failing to start
> your container). Having simple POSTRESQL_FOO=value also means that YOU can
> validate the value and fix it (or default) if it is no longer valid or it
> was renamed.

Yes, but I see no point in that validation.  You say "YOU", but I see
"ME".  I am not going to do the validation when PG does that already
(better and faster way).

Asking again:  Would you be able to implement such variables (redundant
for our images) in your layer?  The image design is so flexible that we
can let this topic solved in your layer, without any clash.

Pavel

**Comment by mfojtik** [comment]
@praiskup if we need to add extra layer on top of images you provide, then what benefit(s) these images provide to us? Yet another way to configure stuff that we will have to support, we don't want and we don't require (so we will most likely disable it...). 

@bparees @danmcp FYI.

**Comment by praiskup** [comment]
@mfojtik wrote:
> @praiskup if we need to add extra layer on top of images you provide, then
> what benefit(s) these images provide to us?

Configuration is not everything, its like .. should I really argue why we
should follow DRY principle at this level?

Whatever new variable 1:1 mapped to PG config-option you will ever invent in
future will be already solved by POSTGRESQL_CONFIG.  So there is no point to
add anything new.

> Yet another way to configure stuff that we will have to support, we don't
> want and we don't require (so we will most likely disable it...).

If you plan to disable `POSTGRESQL_CONFIG` (the generic approach), you should
probably say other reasons than your maintenance efforts - because that is
something already solved in your FROM image.  We maintain that variable
to not limit users ... which you seem to plan to do ;).

The thing is that it is wasting of your resources to (a) invent, (b) maintain,
(c) QA and (d) document each new variable.  All that is solved.  If you dont
really care about a-d (which I do), the only thing you need to add is:

    ```
    $ cat /usr/share/cont-layer/postgresql/postinitdb/openshift.sh
    if test -n "$POSTGRESQL_SHARED_BUFFERS"; do
        # do your validation here, whatever you feel is necessary
        pgcont_config_set_option shared_buffers \
            "$POSTGRESQL_SHARED_BUFFERS"
    fi
    ```

.. and document, maintain and QA that.  All that is redundant for base image
however.

You can do that different way (template your postgresql.conf) or whatever you
feel is the best of course.  There is no clash and I don't see a reason why you
should do completely different images from this perspective.

Pavel

**Comment by hhorak** [comment]
@mfojtik wrote:
> In MySQL, the config file has sections, so I wonder how you will solve that
(section.key=value?).

the only section we care in mysqld daemon is `[mysqld]` -- no issue here

@mfojtik wrote:
> Then you have Tomcat, which is XML config file (pretty ugly one). How is your approach generic to that?

This is not relevant here -- even in tomcat we can use `TOMCAT_CONFIG_somevar=value` or `TOMCAT_CONFIG='somevar=value'` -- the only difference will be invisible for user, which is the implementation of how those values get into the config file. In other words, even @praiskup 's syntax doesn't prevent to implement some intelligent config validator in tomcat.

@mfojtik wrote:
> People use it and they made mistakes (typos, wrong values, etc) and we give them errors that they will then fill in BZ.

There is no difference between getting errors from bash script parsing the config before starting daemon and getting errors from daemon itself during config file parsing. I'm sure that all programs we talk about here are mature enough to report errors in a way users understand it.

Trying to go one step aside and look at this issue from user's PoV -- the only difference between suggested solution is syntax and range of values that may be changed. If we see users are unable to use the `POSTGRESQL_CONFIG='var=value'` way, we can always implement the 'validator'. But let's start with something less complicated for maintainance, images are complex enough even without it.

@mfojtik if the syntax of `POSTGRESQL_CONFIG='var=value'` is hard to be implemented in openshift, please, say it directly.

**Comment by mfojtik** [comment]
@danmcp is this something we want to support in OpenShift?

**Comment by hhorak** [comment]
@mfojtik wrote:
> We should at least try to provide an example or base our users can start
from. Kubernetes is complex tool and having a simply to understand example
in place will be a huge help.
...
Mounting directories with config files is no go for Kubernetes, as we
already discussed that (many times)

So what you're proposing is to actually include the complex master/slave scripts in the image, allow them to be used and thus support them? (since generally what we ship we support)

If we want only to demonstrate that master/slave works, let's create a small layer on top of the supported layer and demonstrate it. But don't include the demo settings into supported image.

**Comment by danmcp** [comment]
@mfojtik @hhorak I don't think we want POSTGRESQL_CONFIG='var=value' where it appends.  I appreciate the attempt to do a catch all but it has too many downsides (doesn't work well for all formats, really needs to be a merge, difficult for UIs to deal with).  Mounting in config is probably a better catch all answer or telling the customer to rebuild a layer with their config changes.

**Comment by praiskup** [comment]
@danmcp wrote:

> @mfojtik @hhorak I don't think we want POSTGRESQL_CONFIG='var=value' where it
> appends.

When the configuration option is not set yet, it appends - yes.  Otherwise it
replaces (and the last wins).  IOW - if you prepare template of
'postgresql-container.conf', you can keep the right order there.

> I appreciate the attempt to do a catch all but it has too many downsides
> (doesn't work well for all formats, really needs to be a merge, difficult
> for UIs to deal with).

Agree it does not match all formats;  but for some of our containers it will
match pretty well.

The 'merge' POV you may decide on per-container basis.  For PG it merges
currently, some may rather append..

Could you elaborate on the UI POV?

> Mounting in config is probably a better catch all answer

+1

> or telling the customer to rebuild a layer with their config changes.

+1.  That will be always possible, though it is really nontrivial additional
work for the user.

**Comment by mfojtik** [comment]
@praiskup  If you going to introduce a configuration mechanism, it should be consistent for all images, correct? So I would like to see an example for JBoss/Tomcat (@goldmann you might be interested in this or jump into this discussion), HTTPD_CONFIG and NGINX_CONFIG. (For example in HTTPD you might define VirtualHost and HTTPD config is crippled XML).

UI: In OpenShift UI we allow users to set the values for the environment variables (or 'parameters' values for Template) which then expands to ContainerManifest Env.
Having POSTGRESQL_CONFIG in place will make it difficult for UI to draw HTML form for users to provide key/values in reasonable way. Also UI will have to know that this ENV var has a 'special' meaning. I can show you this tomorrow during workshop.

So to summarize this, our options are:

* Having supported set of ENV variables that we (GSS) can support and we can document, like `POSTGRESQL_MAX_CONNECTIONS` (which actually set 2 directives in postgresql.conf). You don't want to support this because it introduces 'API' and does not scale (but you haven't told me what you mean with 'scale' yet as we do this approach in v2 that has millions of gears already for couple years).

* `*_CONFIG` which contains a list of key-value pairs that will work well for PG and MySQL but most likely wont work that well for other formats/images and will need more code to deal with those formats that someone will have to support...

* The best option we all agree on is to allow mounting configs files directly, which isn't doable in Kubernetes/OpenShift as it won't scale now (HostDir), there is no way how to define this in Kubernetes and will require to write code/project to do replication of config files across nodes and maintain the security and potentially have performance issues as every container will have to have volume attached.

* Layering is also an option, but most likely not intended for the general audience (more for 'advanced users').

@hhorak @bparees anything I forget ? ^^

**Comment by bparees** [comment]
By scale I think @praiskup means "how much work is it to add 100 configuration options vs 1".  My response to which is:  after several years of running v2 and responding to user requests to add config options, we support maybe 10 or so settings for most cartridges(images) and it's not that much work to add one.

the other problem with volumes (as i stated earlier) is that providing configuration via a volume mount pretty much violates the fundamental concept of containers being self-contained app+config.  (yes injecting environment variable config also violates that, but at least in that case it's easy to ship a script that starts the image w/ the right injected env, or use openshift which provides templates for doing so).  Telling people to use volumes leaves them on their own to design a way to manage those volumes, as @mfojtik alludes to.

Lastly, I see a lot of comments about how more intelligence could be wrappered into the handling of these environment variables under the covers...but as soon as you start doing that, i'm not seeing how you're better off in terms having a less work/more scalable approach.  You're basically back to where we're starting.

**Comment by praiskup** [comment]
> @praiskup If you going to introduce a configuration mechanism, it should be
> consistent for all images, correct? So I would like to see an example for
> JBoss/Tomcat (@goldmann you might be interested in this or jump into this
> discussion), HTTPD_CONFIG and NGINX_CONFIG. (For example in HTTPD you might
> define VirtualHost and HTTPD config is crippled XML).

Well, it would be fine to have the same approach for all our images, +1.  But it
does not seem to be realistic requirement.  I don't think its a reason to stop
using that approach if that works for at least some containers.  You said you
will disable it in your layer, if you feel this is positive in *any* way,
you are always free to do that.

> UI: In OpenShift UI we allow users to set the values for the environment
> variables (or 'parameters' values for Template) which then expands to
> ContainerManifest Env.

Example of the Manifest Env somewhere?  Would be useful.

> Having POSTGRESQL_CONFIG in place will make it difficult for UI to draw HTML
> form for users to provide key/values in reasonable way.

It does not sound like show-stopper, its a presentation.  Nothing that
must be done now, nothing really complicated.  I don't see the difference
between your and the generic approach.

> Also UI will have to know that this ENV var has a 'special'
> meaning. I can show you this tomorrow during workshop.

Not everybody from github will be able to see that workshop.

> So to summarize this, our options are:
> 
>     Having supported set of ENV variables that we (GSS) can support and
>     we can document, like POSTGRESQL_MAX_CONNECTIONS (which actually set
>     2 directives in postgresql.conf). You don't want to support this
>     because it introduces 'API' and does not scale (but you haven't told
>     me what you mean with 'scale' yet as we do this approach in v2 that
>     has millions of gears already for couple years).

I told that already, why it will not scale.  You have two variables now,
only, fine.  In OpenShift, there are probably customers not asking for
special customization, which is really weird..  Once you will try to think
about master/slave, add things like fail over etc..  You'll need more.

I don't reject this idea for OpenShift only, I just argue its wasting of
(your) resources.  I am not able to add and document and test and..  any
variable other (non-OpenShift) customers will want, if there is no generic
way to do that.  Sorry, but *this* will not scale.

@mfojtik, if you claim its not wasting of resources, why you reject that
idea of supporting those variables yourself in your layer?  If you are
really convinced you don't need more variables, I've made it terribly easy
to achieve your goal.  Its just about one additional shell script, where
is the problem?

>     *_CONFIG which contains a list of key-value pairs that will work
>     well for PG and MySQL but most likely wont work that well for other
>     formats/images and will need more code to deal with those formats
>     that someone will have to support...

Yes, I'm not going to support that formats hopefully.  Somebody else
interested must comment how far he will be able to go with environment
variable bloat, maybe @hanzz could comment httpd case.

>     The best option we all agree on is to allow mounting configs files
>     directly, which isn't doable in Kubernetes/OpenShift as it won't
>     scale now (HostDir), there is no way how to define this in
>     Kubernetes and will require to write code/project to do replication
>     of config files across nodes and maintain the security and
>     potentially have performance issues as every container will have to
>     have volume attached.

How about to make this happen?  If we all agree that it is the best
option, yesterday was too late.

@mfojtik, for the beginning the terribly simple approach:  why you can't
prepackage your configuration on all your hosts consistently, similarly
your shell callbacks we will call for you?

>     Layering is also an option, but most likely not intended for the
>     general audience (more for 'advanced users').

Agree.

**Comment by praiskup** [comment]
> By scale I think @praiskup means "how much work is it to add 100
> configuration options vs 1". My response to which is: after several
> years of running v2 and responding to user requests to add config
> options, we support maybe 10 or so settings for most cartridges(images)
> and it's not that much work to add one.

Once we release _general_ non-openshift specific images, the requirements
will be different probably.  I mentioned it before, its weird there are
not other requirements for configuration;  the amounts of configuration
options exists for some reason probably and I am asking why users in
OpenShift don't need them?

> the other problem with volumes (as i stated earlier) is that providing
> configuration via a volume mount pretty much violates the fundamental
> concept of containers being self-contained app+config. (yes injecting
> environment variable config also violates that, but at least in that
> case it's easy to ship a script that starts the image w/ the right
> injected env, or use openshift which provides templates for doing so).
> Telling people to use volumes leaves them on their own to design a way
> to manage those volumes, as @mfojtik alludes to.

I can pretty much agree here.  Thats why I don't reject the environment
variable idea completely (and why I implemented POSTGRESQL_CONFIG).  Just
calling the 'volume' approach preferred.

> Lastly, I see a lot of comments about how more intelligence could be
> wrappered into the handling of these environment variables under the
> covers...but as soon as you start doing that, i'm not seeing how you're
> better off in terms having a less work/more scalable approach. You're
> basically back to where we're starting.

I'm not sure if I get this point correctly.  But for PostgreSQL images Iwill be
able to avoid respins for customers' requests to add new config variable.
That is huge help I can solve *now*.

**Comment by mfojtik** [comment]
On May 20, 2015 10:25 PM, "praiskup" <notifications@github.com> wrote:
>>
>> @praiskup If you going to introduce a configuration mechanism, it should
be
>> consistent for all images, correct? So I would like to see an example for
>> JBoss/Tomcat (@goldmann you might be interested in this or jump into this
>> discussion), HTTPD_CONFIG and NGINX_CONFIG. (For example in HTTPD you
might
>> define VirtualHost and HTTPD config is crippled XML).
>
> Well, it would be fine to have the same approach for all our images, +1.
But it
> does not seem to be realistic requirement. I don't think its a reason to
stop
> using that approach if that works for at least some containers. You said
you

So you don't care about consistency anymore?

> will disable it in your layer, if you feel this is positive in any way,
> you are always free to do that.
>>
>> UI: In OpenShift UI we allow users to set the values for the environment
>> variables (or 'parameters' values for Template) which then expands to
>> ContainerManifest Env.
>
> Example of the Manifest Env somewhere? Would be useful.

https://github.com/openshift/origin/blob/master/examples/sample-app/application-template-stibuild.json#L120

>>
>> Having POSTGRESQL_CONFIG in place will make it difficult for UI to draw
HTML
>> form for users to provide key/values in reasonable way.
>
> It does not sound like show-stopper, its a presentation. Nothing that
> must be done now, nothing really complicated. I don't see the difference
> between your and the generic approach.

It must be done now as we have release soon. You will see if you check the
sample app and try to run it yourself. I am sad that valid point are
ignored here and valid arguments rejected.

>>
>> Also UI will have to know that this ENV var has a 'special'
>> meaning. I can show you this tomorrow during workshop.
>
> Not everybody from github will be able to see that workshop.

I am pretty much explained the problem here already.

>>
>> So to summarize this, our options are:
>>
>> Having supported set of ENV variables that we (GSS) can support and
>> we can document, like POSTGRESQL_MAX_CONNECTIONS (which actually set
>> 2 directives in postgresql.conf). You don't want to support this
>> because it introduces 'API' and does not scale (but you haven't told
>> me what you mean with 'scale' yet as we do this approach in v2 that
>> has millions of gears already for couple years).
>
> I told that already, why it will not scale. You have two variables now,
> only, fine. In OpenShift, there are probably customers not asking for
> special customization, which is really weird.. Once you will try to think
> about master/slave, add things like fail over etc.. You'll need more.

You already refused master/slave.

>
> I don't reject this idea for OpenShift only, I just argue its wasting of
> (your) resources. I am not able to add and document and test and.. any

You are, we do it already.

> variable other (non-OpenShift) customers will want, if there is no generic
> way to do that. Sorry, but this will not scale.

Still haven't heard argument that proves it.

>
> @mfojtik, if you claim its not wasting of resources, why you reject that
> idea of supporting those variables yourself in your layer? If you are
> really convinced you don't need more variables, I've made it terribly easy
> to achieve your goal. Its just about one additional shell script, where
> is the problem?
>>
>> *_CONFIG which contains a list of key-value pairs that will work
>> well for PG and MySQL but most likely wont work that well for other
>> formats/images and will need more code to deal with those formats
>> that someone will have to support...
>
> Yes, I'm not going to support that formats hopefully. Somebody else
> interested must comment how far he will be able to go with environment
> variable bloat, maybe @hanzz could comment httpd case.
>>
>> The best option we all agree on is to allow mounting configs files
>> directly, which isn't doable in Kubernetes/OpenShift as it won't
>> scale now (HostDir), there is no way how to define this in
>> Kubernetes and will require to write code/project to do replication
>> of config files across nodes and maintain the security and
>> potentially have performance issues as every container will have to
>> have volume attached.
>
> How about to make this happen? If we all agree that it is the best
> option, yesterday was too late.

It is the best option for Docker not for Kubernetes.

**Comment by praiskup** [comment]
> > variable other (non-OpenShift) customers will want, if there is no generic
> > way to do that. Sorry, but this will not scale.
>
> Still haven't heard argument that proves it.

You obviously don't expect other will come (which would prove it).  We do not
want make openshift our only customer, or do we?

> > I don't reject this idea for OpenShift only, I just argue its wasting of
> > (your) resources. I am not able to add and document and test and.. any
>
> You are, we do it already.

When you are doing all that steps, the following unanswered paragraph is still
valid:

**Comment by hhorak** [comment]
@praiskup wrote:
> But for PostgreSQL images Iwill be able to avoid respins for customers' requests to add new config variable.
That is huge help I can solve now.

+1 I think this is very important note, we cannot expect docker images will be updated just for adding new configuration. That would be really wasting of resources.

@mfojtik Also it seems you have an opinion that a few specific options that may be changed in the config file is a benefit for users, is that correct? I see it may be valid for openshift UI (to offer only few options), but I really can't see how this will make the docker image more valuable for users.

Anyway, one of the arguments above was that the set of options users will need is already known, based on experiences from v2. But the docker images we're creating now will run in much more different environments, so I can imagine to tune up the image the same as users tune up database usually -- e.g. increase buffer sizes, which is indeed a legitimate request. Actually, from my PoV it is quite strong requirement I have for the mysql image (based on the knowledge how mysql database is used). Plus, there are many configuration options that add features users will need in docker images the same as they need them in rpm -- ssl settings, slow_query_log_file, optimizer_switch and more.. Use cases for using those options don't get changed by the way the daemon is run -- whether as standalone process or as container in cloud.

**Comment by praiskup** [comment]
@mfojtik wrote:
> It must be done now as we have release soon. You will see if you check the
> sample app and try to run it yourself.

I get this very important point.  But why format of presentation of some
variable in UI would block your release?  Anyway, what changes are acceptable
right now during pre-release time?  I mean in your current Dockerfiles and images.

> I am sad that valid point are ignored here and valid arguments rejected.

Seconded :(, I feel ignored all the time.

Maintainers of other tools should be in the loop asap, we need to make both
parties happy.

But speaking about PostgreSQL - is there problem with the approach I propose -
because I don't see any.  You can trivially add something your UI needs, and if
you really need to make POSTGRESQL_CONFIG disabled, you can do that easily -
too.  Anything else which blocks us here, to make PG solved?

**Comment by mfojtik** [comment]
On May 20, 2015 11:17 PM, "praiskup" <notifications@github.com> wrote:
>
> @mfojtik wrote:
>>
>> It must be done now as we have release soon. You will see if you check
the
>> sample app and try to run it yourself.
>
> I get this very important point. But why format of presentation of some
> variable in UI would block your release?

Because the variable is important for us. And the presentation is also
important as that will be the way people will interact with the image
settings in OpenShift UI.

Anyway, what changes are acceptable
> right now during pre-release time? I mean in your current Dockerfiles and
images.

Any change that won't break our QA or functionality like initialization,
services, scale scenarios.

>>
>> I am sad that valid point are ignored here and valid arguments rejected.
>
> Seconded :(, I feel ignored all the time.
>
> Maintainers of other tools should be in the loop asap, we need to make
both
> parties happy.
>
> But speaking about PostgreSQL - is there problem with the approach I
propose -
> because I don't see any. You can trivially add something your UI needs,
and if
> you really need to make POSTGRESQL_CONFIG disabled, you can do that
easily -
> too. Anything else which blocks us here, to make PG solved?

The current plan is that we either consume your images as they are, without
rebuilding them to disable/add out stuff or we go with our current set of
images. @danmcp correct me if I am wrong.

**Comment by praiskup** [comment]
> The current plan is that we either consume your images as they are, without
> rebuilding them to disable/add out stuff or we go with our current set of
> images. @danmcp correct me if I am wrong.

Is this pre-release plan only, or is this long term status?

**Comment by danmcp** [comment]
@praiskup The plan/goal is that we can ship one set of images that satisfy all our goals.  We want this because:

- We don't want to have to maintain multiple images
   - Layering the images will also be long term work that's difficult to track.  Every time something is added at a lower layer that isn't desired at an upper layer there will be work to disable.  Plus there is no way to automatically detect those sorts of changes unless something breaks.
- Everyone gets a consistent experience regardless of what platform they are using the images on
   - Apps built on images used outside of OpenShift can be easily be transitioned to being used inside of OpenShift and vice versa.

**Comment by praiskup** [comment]
Ideally, but we need to take into account that the images we will provide
are not only for OpenShift - and thus these need to stay general.
OpenShift has specific requirements, which will not be useful by other
users.

To think about it a from the other side - you will dislike any change
requested by different customers probably - because that is not something
you really need and you refuse to support it.  Thats completely fine -
because we can't just bloat the generic image with something customer
specific.

We are not 100% compatible, but about in 95% - and it still sounds it is
worth cooperation.  I believe there would be tons of problems you'll be
fine to delegate to FROM image.

**Comment by danmcp** [comment]
> Ideally, but we need to take into account that the images we will provide
are not only for OpenShift - and thus these need to stay general.

Agreed

> OpenShift has specific requirements, which will not be useful by other
users.

What's this list?

> To think about it a from the other side - you will dislike any change
requested by different customers probably - because that is not something
you really need and you refuse to support it.  Thats completely fine -
because we can't just bloat the generic image with something single-customer
specific.

We generally take the approach of covering the 80% case.  The other 20% will need to build custom images.

> We are not 100% compatible, but about in 95% - and it still sounds it is
worth cooperation.  I believe there would be tons of problems you'll be
fine to delegate to FROM image.

I am not quite sure what you are saying here.  I don't think we want to FROM any intermediate images for the reasons I listed previously.

**Comment by hhorak** [comment]
@danmcp wrote:
>>OpenShift has specific requirements, which will not be useful by other users.

> What's this list?

The environment variables syntax, which is what this whole thread is mainly about. 7 messages above I gave you specific reasons why the few-specific-options way won't fit the needs of non-openshift users.

> We generally take the approach of covering the 80% case. The other 20% will need to build custom images.

But I still don't see why we should deliberately support just 80% if we can quite easily support 95% while saving maintenance cost at the same time.

**Comment by mfojtik** [comment]
@hhorak the problem with your approach is:

* Complicate UI(s) flows
* It is inconsistent across images (want to confirm this from Tomcat, Nginx and httpd folks)
* Require to invent and maintain a tooling that provide merging
* Require to add more scripting to deal with other formats
* It's basically out-source all problems with Docker images to the software running on image
* It is complex for the target audience that will be consuming this (regular users, without knowledge of Postgresql, Mysql and without administration/bootstrapping experience)

The problem with our approach as I can see from your comments are:

* It won't scale (which isn't true as we can prove it scale well in OpenShift v2)
* It require you to add, document and maintain couple env vars and add more if customers request them (is that bad actually? having customers using your images and requesting more features?)
* It introduce API (is that bad as well?)

I think we are moving in circles here, if you don't want to provide features we require then said it and we can move on in OpenShift.

**Comment by praiskup** [comment]
> It won't scale (which isn't true as we can prove it scale well in OpenShift
> v2)

Sorry, but if you require only two configuration options, its like toy example I
must say.  Its probably enough to drive wordpress in Openshift, but that does
not prove that PostgreSQL should drop all the features it has.  Neither should
we refuse to support it.

> It require you to add, document and maintain couple env vars and add more if
> customers request them (is that bad actually? having customers using your
> images and requesting more features?)

Is it worth doing if we can solve it easily and forever?  PostgreSQL implements,
maintains and documents all the options for us.  I can't duplicate that work.

> It introduce API (is that bad as well?)

Wrong API matching OpenShift needs, thats bad.

**Comment by praiskup** [comment]
> Wrong API matching OpenShift needs, thats bad.

I mean OpenShift *only* needs of course.

**Comment by mfojtik** [comment]
> Sorry, but if you require only two configuration options, its like toy example I
must say. Its probably enough to drive wordpress in Openshift, but that does
not prove that PostgreSQL should drop all the features it has. Neither should
we refuse to support it.

Nobody is saying you will drop any features as you can always put layer on top of your image with config file which is what **your** customers will do. The ENV configuration will provide a way to configure the image behavior for users that **don't** understand postgresql and they **don't** want to configure everything, just fields that are common use cases. 

> Is it worth doing if we can solve it easily and forever? PostgreSQL implements,
maintains and documents all the options for us. I can't duplicate that work.

Read above....

**Comment by mfojtik** [comment]
@praiskup do you have **any** use cases from PM to support POSTGRESQL_CONFIG at all? Or is that something you made up for us? If so, I'm saying we don't want it and our users don't want it either. Reasons explained here: https://github.com/sclorg/rhscl2dockerfile/issues/17#issuecomment-104183798

**Comment by praiskup** [comment]
> Nobody is saying you will drop any features as you can always put layer on top
> of your image with config file which is what your customers will do.

This is false argument.  Even YOU refuse to do it - and you have huge amount of users
doing the same things which makes it really worth doing additional layer.  If you feel
this is not doable approach, how can we require others to do it.

**Comment by mfojtik** [comment]
@praiskup well, your argument was that the best option is to mount config files... what prevents it when you do the ENV config we require? Nothing. Your users will still be able to do it. Then, what we are talking about here?

**Comment by mfojtik** [comment]
@praiskup also on last meeting, your PM mentioned that your customers will go customize the images you will provide and build on top of them, which indicate that your use-case is what I described above.

**Comment by mfojtik** [comment]
@praiskup just talked to @goldmann, they do have at least 10 environment variable for now to use with their Docker images (Jboss). So they are following what we do now. For 3.1+ release they are thinking about creating universal configuration tool you guys should collaborate on that with them.

**Comment by goldmann** [comment]
Yah, like @mfojtik said - we need some general configuration tool, because current bash scripts are hard to maintain and we're unable to test it. Image like EAP or Tomcat need to be pretty customized on each boot, but the way we describe what needs to be changed needs to be in an easy format. Additionally everything needs to be testable. Bash scripts are not. I'll have more details on our idea soon.

**Comment by goldmann** [comment]
One thing - this is a more long-term idea (probably OS 3.1), for 3.0 we (JBoss) will stick with the scripts we have today.

**Comment by praiskup** [comment]
@goldmann, you need to think about the support of environment variables even in
3.1, however.  But yes, general configuration mechanism is prefect and I support
that idea.  But count with that it will be disabled for OpenShift - because that
is *not* something OpenShift wants.  We have something like that solved for
some of our images and it *is* problem.

**Comment by mfojtik** [comment]
@praiskup @goldmann approach is general for all images :-)

**Comment by praiskup** [comment]
@mfojtik, then its fine to support both approaches for me in future.

**Comment by praiskup** [comment]
@mfojtik, you'll have to disable both general approaches in future, however.  Based on
your current statements - you do not want them, yes?  Moreover - all your environment
variables will be already there and you will need to support that.  Thats what I try to avoid
right now.

**Comment by goldmann** [comment]
@praiskup

I think you got me wrong - we'll still act based on the env variables defined - this will be source of the values and information what needs to be done within the enabled modules. (A module is a set of available configuration changes for specifc use case: think all possible changes that are required to run JBoss EAP on OpenShift for example).

OpenShift is not that involved in the tool. This tool should be self-contained (backed into image) and probably be the entrypoint (`CMD`).

Like I said, we'll try to prepare something we can collaborate soon and I'll let you guys know.

**Comment by praiskup** [comment]
@praiskup

> I think you got me wrong - we'll still act based on the env variables
> defined - this will be source of the values and information what needs to
> be done within the enabled modules.

So for PostgreSQL - still one VARIABLE for one configuration option?

> (A module is a set of available configuration changes for specifc use
> case: think all possible changes that are required to run JBoss EAP on
> OpenShift for example).

Perfect!

> OpenShift is not that involved in the tool. This tool should be
> self-contained (backed into image) and probably be the entrypoint (CMD).

Something like atomic is right now?

> Like I said, we'll try to prepare something we can collaborate soon and
> I'll let you guys know.

Please, do it rather sooner than later.

In the meantime, do you see any reason why I should reject POSTGRESQL_CONFIG
for PostgreSQL container in fedora/centos/rhel if that simply solves that for
this particular container?

**Comment by goldmann** [comment]
One or more variables - we don't care. It's up to the image developer, which has in control both sides of the problem:

1. The image and what env variables that are generated/required,
2. The configuration tool modules that react based on the env variables.

My personal preference is of course to have many env variables. Do not need to parse stuff, which can break at any time.

My advice here is: use whatever fits you the best for now, since you are the owner and you will support it. Other thing is how it'll be implemented/exposed in OS. This is a part where you need to look at what's possible in OS, or what OS team can do for you.

In general - if you go with multiple env vars and think about it carefully - after the change to the new tool - it should still wok the same way. This is because we're talking here about two things:

1. How to instruct the container (env vars)
2. How to translate it into configuration changes.

The tool I'm talking about is only about 2nd issue.

**Comment by praiskup** [comment]
@goldmann, cool, thanks for the info!

**Comment by hanzz** [comment]
For httpd (not talking about rhscl now, but generally about some future httpd integration with docker/kubernetes), I'm playing with an idea of using shared volume for configs in docker world and Kubernetes annotations in Kubernetes world. Can't say if something like Kubernetes annotations exist in Openshift.

Basically if I imagine the use-cases for httpd in container, I end up with something like:
1) serve static data on domain.tld/something
2) proxy domain.tld/something to another application
3) be a base image for another application. In this case, I think people will just base their app on the image and inject the configuration files themselves, so in this case, there's no need for configuration.

The difference between httpd and other services is that it should, in my opinion, be possible to reconfigure it even when it's already running. You might want to add another vhost/web-app/static-data directory later and it's not practical to restart the containers because of that.

Therefore I think env variables are not so ideal for this use-case, because once you start the container configured using env vars, it's just hard to change it. I like the idea that the container is monitoring the configuration stored "somewhere" (shared volume, kubernetes annotations) and reloads itself when the configuration changes.

My current approach (again, not in rhscl) is based on JSON format describing some really basic stuff which is later translated into httpd configuration (Imagine it like "proxy domain.tld/blog to http://IP_of_another_container/wordpress", "serve this volume as domain2.tld/static" and so on.). I use independent configuration language (JSON) for the basic stuff, because later it could be possible to use the same configuration format even for nginx (one just have to write translation tool to translate that JSON into nginx config file). However, admin can still inject raw config file if he wants something special.

I'm writing it here just to let other think about this way of handling configurations. I'm not saying this is optimal solution, because I'm still new to Kubernetes/Docker, so I'm more than happy for others to correct me :).

**Comment by danmcp** [comment]
@hhorak 

> The environment variables syntax, which is what this whole thread is mainly about. 7 messages above I gave you specific reasons why the few-specific-options way won't fit the needs of non-openshift users.


@mfojtik mostly answered this but your statement was OpenShift requirements that aren't useful to other users.  I don't think env vars fits into that list.  I think your argument against them is it isn't a complete solution (none of them are yet).  But I don't think there is an argument that exposing individual variables isn't useful for the non OpenShift cases.

> But I still don't see why we should deliberately support just 80% if we can quite easily support 95% while saving maintenance cost at the same time. By 'maintenance cost' I mean adding new config variables, that maybe wouldn't come from openshift users, but would definitely come from non-openshift users, that will need to set up the container a bit more.

Accepting the percentages stated, the choices seem to be roughly:
- Cover the 80% with a consistent experience
- For some use cases cover the 95%

I won't list as the flaws of each (they are already covered enough).  But the inconsistency of the "95% " case is its biggest flaw to me.  And you end up with 4 or more options because you still haven't covered the 100% case:

- Individual env vars
- POSTGRESQL_CONFIG style
- mounting in config
- rebuilding images


@praiskup 

> @mfojtik, then its fine to support both approaches for me in future.

I actually agree with this general mindset.  I am fine with having a simple solution and a catch all solution.  These cases for example are those:

1) Individual env vars (Simple but not a catch all)
2) mounting in config (Catch all 1 but means you might have to replace a lot of config)
3) rebuilding images (Catch all 2 but means you have to build stuff yourself)

Maybe there is room for a 1a).  But I don't think that implementation is what we are talking about for POSTGRESQL_CONFIG because it's making too many assumptions and has too many problems:

- Append will always work.  Even if it does work it's not very user friendly to read after
- Doesn't work well with UIs.  Both hard to implement and doesn't look great
- It's not consistent across images (doesn't work for xml, yaml, json, etc)
...

And maybe if 1a) is good enough I would even vote to get rid of 1).

**Comment by hhorak** [comment]
Since we've got stuck here a bit, we had off-github discussion with @mfojtik about how to solve this issue, the answer is: we'll ship an `openshift` binary (bash script) in the RHSCL images that will add everything what doesn't seem to be much usable in non-openshift workflow (parsing the 'Individual env vars' for example). This `openshift` binary will be then used as `CMD` in openshift (thus should be on $PATH). The binary will do the same as default `CMD` (whatever it is for particular image) but could do something more specific.

This seems to be acceptable compromise, since as openshift approaches the release, they're not able to change anything in API (environment variable names for configuration adjustments for example). But since we still want to have only one set of docker images, not two sets, we need to think about some compromises...

**Comment by hhorak** [comment]
One another small but quite important thing -- since it is planned to have some better way to configure containers in the future (as mentioned above) and since the env. variables to configure deamons should be used only through web UI in the openshift -- we shouldn't document them in the image and state it is not recommended to use them outside of openshift. That way we can remove them much easier as soon as we have the proper configuration tool.

**Comment by hhorak** [comment]
@mfojtik either it's because it's late already or my head doesn't work anymore, but after many back and forth with the master/slave, I'm not sure anymore where we are now.. Is master/slave for mysql and mongodb (there is nothing like that in postgresql) a requirement and you can't use RHSCL images without it or not? I think it is not a requirement as for today, but rather would like to have a confirmation here from you, because it is not clear from the above.

**Comment by bparees** [comment]
@hhorak it's not a requirement today, i'd expect it to be in the near future if we want to provide some value here.

**Comment by praiskup** [comment]
@hanzz wrote:
> The difference between httpd and other services is that it should, in my
> opinion, be possible to reconfigure it even when it's already running.

I'm not sure httpd is special from this POV.  Any service has this problem.

**Comment by mfojtik** [comment]
@hhorak @praiskup containers are (at least in Kubernetes) stateless in terms of their configuration. The configuration must be persisted in the Docker image and you should **never** reconfigure a running container there. Typically, you will not run just one container but N containers and you will have copy that change to every of those. Also containers might be destroyed or moved to another node (which means destroy/start) at any time and in that case your change will be lost or you will break the replication.

Reconfiguring containers might have sense when you run them locally and you want to commit the configuration as a new image, but that is possible even now (docker exec, make change, docker commit), so no special tooling is needed around that IMHO.

**Comment by mfojtik** [comment]
hups the above was ment to @hanzz, but might be valuable for Honza as well ;-)

**Comment by hanzz** [comment]
@mfojtik wrote
> Typically, you will not run just one container but N containers and you will have copy that change to every of those and it is considered as very bad practice.

I'm talking about case when you want to configure "httpd" container for the "wordpress" container. In that case, it is possible to just store the configuration in the "wordpress" kubernetes service definition and httpd container can pick that up automatically.

The configuration of httpd container changes with the presence of other Kubernetes services. I don't see why this is considered as a bad practice. I don't have to copy anything anywhere. IMHO this is used by other projects like skyDNS too (it stores the configuration in etcd and you configure running skyDNS instance by pushing new configuration into etcd, skyDNS picks that up automatically).

> Also containers might be destroyed or migrated to another node (which means destroy/start) at any time and in that case your change will be lost or you will break the replication.

They won't be lost. If you migrate httpd container to another host, it will just reload the configuration from the Kubernetes API. If you migrate the web-app (wordpress or something like that), nothing will change in httpd container unless you remove the wordpress service. In case you remove the wordpress service, clients cannot use the wordpress, so httpd will remove wordpress configuration and reload automatically.

Note that this is only valid for the services like httpd which can be used in cooperation with other services running in the cloud, therefore i wrote it's somehow special to httpd use-case.

**Comment by hanzz** [comment]
You can of course store the wordpress related configuration in httpd container using ENV variables, but why would you do that? Typically wordpress related configuration always belonged to the wordpress RPM/service and so on.

If you do it another way, you basically disallow web-app developers to create self-contained web-app image working with httpd out-of-box, because people will always have to configure httpd manually (add the correct ENV vars there).

Using the shared configuration, web-app developer can have the httpd configuration embedded in the Kubernetes service file definition and user can just do "atomicapp run wordpress:fpm" or something like that and if he has httpd installed (And what's the more important - if he has any webserver which can handle that shared configuration), it will just work. That case, it is possible to write the nginx parser for the shared configuration and you can use nginx. Or even openshift can have its own thing doing the proxying based on that configuration.

**Comment by mfojtik** [comment]
@hanzz 

> I'm talking about case when you want to configure "httpd" container for the "wordpress" container. In that case, it is possible to just store the configuration in the "wordpress" kubernetes service definition and httpd container can pick that up automatically.

For this case you have STI project and the PHP image. You define BuildConfig 'Source' strategy and you provide the GIT URL of your wordpresssource code. Then the final Docker image is composed from the Wordpress source code and the 'builder' image here. In this case, the builder image will be 'php + http'. You can configure the 'httpd' (or nginx) during the build, inside 'assemble' script.

> You can of course store the wordpress related configuration in httpd container using ENV variables, but why would you do that? Typically wordpress related configuration always belonged to the wordpress RPM/service and so on.

The Wordpress configuration is persisted in the Docker image. It is not persisted in ENV var or Kubernetes Service (btw. k8s services does not allow that anyway).

> If you do it another way, you basically disallow web-app developers to create self-contained web-app image working with httpd out-of-box, because people will always have to configure httpd manually (add the correct ENV vars there).

They will not have to configure anything manually, the 'php' builder image will be based on httpd image and allow configuration of httpd during the build time.
Actually (@bparees, @danmcp correct me if i'm wrong) we will most likely have special 'wordpress' image that will be tuned to run wordpress.

> Using the shared configuration, web-app developer can have the httpd configuration embedded in the Kubernetes service file definition

AFAIK you can't do that and k8s service is not meant to provide configuation to containers, but to group multiple Pods together. Citation:

> A Kubernetes Service is an abstraction which defines a logical set of Pods and a policy by which to access them. https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/services.md()

**Comment by hanzz** [comment]
@mfojtik 
> For this case you have STI project and the PHP image. You define BuildConfig 'Source' strategy and you provide the GIT URL of your wordpresssource code. Then the final Docker image is composed from the Wordpress source code and the 'builder' image here. In this case, the builder image will be 'php + http'. You can configure the 'httpd' (or nginx) during the build, inside 'assemble' script.

I think I'm now starting to see your points, because I always thought we will distribute Docker images, while you plan the users to build the images themselves (using some openshift automation if I got it right).

> The Wordpress configuration is persisted in the Docker image.

If that image is intended to be distributed, then I think this is not right. I always thought people will distribute the images, not the Dockerfiles. If that presumption is right, then how will you define that you want "wordpress running on "ww.domain.tld/blog" when that configuration is already stored in the built image?

If the users are going to build the image themselves, they will have to rebuild it when changing the configuration, right? (Sorry if that's dumb question)

> It is not persisted in ENV var or Kubernetes Service (btw. k8s services does not allow that anyway).

Technically kubernetes does allow that. See [Kubernetes annotations](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/annotations.md).

I want to especially quote the last sentence from it:
>> Yes, this information could be stored in an external database or directory, but that would make it much harder to produce shared client libraries and tools for deployment, management, introspection, etc.

> They will not have to configure anything manually, the 'php' builder image will be based on httpd image and allow configuration of httpd during the build time.

Fair enough, I've just thought people will deploy already built images. If you are going to build the images automatically from the "source-code", then this makes sense. I'm trying to address the case when the image is already built and you want to use it.

> Actually (@bparees, @danmcp correct me if i'm wrong) we will most likely have special 'wordpress' image that will be tuned to run wordpress.

Yeah, I was taking wordpress only as an example of web-app, but I was talking about web-apps as general.

> AFAIK you can't do that and k8s service is not meant to provide configuation to containers, but to group multiple Pods together. Citation:

See my link to Kubernetes annotations above.

Btw, does the atomicapp people know about the way how you are going to ship the images, because if I'm right, they now presume you will use already-built docker images?

**Comment by hanzz** [comment]
Hm, the citation of the "Kubernetes annotation" quote is messed up in the post above, but I'm sure you will parse it as expected ;).

**Comment by hhorak** [comment]
@hanzz @mfojtik I've gone through the discussion and it kind of seems we have a way to go, don't we?

What about the issue about port and non-root user for all httpd processes (including the main one), which is also openshift's quite important request? It was supposed to be tracked in https://github.com/sclorg/rhscl2dockerfile/issues/14, but main discussion happened at https://github.com/sclorg/rhscl2dockerfile/pull/10 so far. Do we have a solution for that without loosing any security (there is the separation issue when storing all logs under one user)..?

**Comment by hanzz** [comment]
@mfojtik 
We are not against switch from 80 to 8080.

Talking about non-root, I would like some clarification.

- As I understood earlier when reading the #10, you want to run httpd (and other services) as non-root to grant them permissions to access just their own container/volume (or maybe set of containers/volumes). I was trying to find out, how do you parametrize that UID. In the sti-base, there is user "default" with uid 1001. I don't see how that can be changed per container. Does that mean that all STI containers share the same UID? I see your point that it's better than sharing "root", but I thought it will work like someone described in #10:

> Random(assigned) uids ensure you can setup the volume such that exactly the containers that you want to have access, can access it.

- The big problem we have with httpd as non-root is that httpd logs would be created by the same non-root user as the httpd is running as. This brings possibility that someone hacking the httpd can remove/change the logs and this could be considered as security issue. I presume that you want to use httpd24 as a base image for other web-based applications, so in your case, it actually could make a sense to log error_log to stderr, because the error_log would contain just log messages produced by that single application. Right now we are still discussing, if that's the primary use-case for httpd24 image, or we expect people to use it also in classic way when the logging to stderr would be quite useless.

- Another issue with non-root httpd is that, according to [sti-php](https://github.com/openshift/sti-php/blob/master/5.5/Dockerfile.rhel7#L40), it would need /etc/httpd write-able by the "default" httpd user. That's really problem, because hacked httpd could be used to change its own configuration. Maybe that's bug or hackish workaround in the sti-php, because currently I don't know why you would need that...

It should also be noted that httpd24 docker image is starting to be whole new product based on the httpd24 SCL RPM IMHO and should be treated like that. With the changes like those above, it needs to be retested and some security things have to be evaluated, which needs some time and cannot be done over the night.

**Comment by mfojtik** [comment]
@hanzz I think @csrwng will provide you with better answer as he knows more about random UIDs than me ;-)

**Comment by hhorak** [comment]
@mfojtik wrote:
> @hanzz I think @csrwng will provide you with better answer as he knows more about random UIDs than me ;-)

That issue is tracked in https://github.com/sclorg/rhscl2dockerfile/issues/18 btw.

**Comment by hanzz** [comment]
@hhorak Thanks, I will read that one too (lot of reading.... :) )

**Comment by l-d-j** [comment]
Hi, I've with @goldmann started a container configuration tool project. It could serve a as a better tool for configuring stuff in containers in both build and run time. Its designed to have plugin based architecture, so you can easily extend the tool to be able to suite your purposes. Its still in prototyping state and you can see it here:
https://github.com/containers-tools/cct
also there is some of concept and samples described in a wiki:
https://github.com/containers-tools/cct/wiki
Be aware that tool is in the early hacking state (started a few days ago) so it just can run some early dummy, shell and xml module (see README.md). Any feedback is welcome and I will try to address any comments/issues/suggestion to make tool really usable. I'm also working now on better parameter passing to modules and automatic environment variables usage in them. I plan to have the modules operation and parameters passing in a final state during next week, I will officially announce the tool during next week with a proper doc and some usable modules.
cc @vpavlin @mfojtik @praiskup

**Comment by praiskup** [comment]
@l-d-j thanks for the info!  Sounds promising, if I get it right - its
something like lightweight augeas for configuration, we could
try to implement support for it in https://github.com/devexp-db/cont-lib .

That might be highly usable for containers which don't care the python will be
also installed into container.

**Comment by goldmann** [comment]
Please skip for now the language in which the PoC was written. Let's focus on the functionality.

**Comment by hhorak** [comment]
@l-d-j Reading the current version of the doc, I still have many questions on my mind..yaml file containing instructions

**Comment by goldmann** [comment]
Yep, we choose YAML because of two reasons:

1. It's simple to read by human
2. It's already proven to do the job (hint: Ansible)

We need to move from bash script thinking. The module that has support for bash execution here was the simplest to implement and this is what @l-d-j did. We are **not** going to rewrite our scripts in bash commands and put them in YAML. Our goal is to create specialized modules (eap, tomcat, etc) and execute them parmetrized.

OK, let's move to the [cct tool issue tracker](https://github.com/containers-tools/cct/issues) for anything that it relevant. This issue is already talking about everything and nothing at the same time :)

**Comment by l-d-j** [comment]
@praiskup thanks for great feedback. yep its inspired by augeas, salt, ansible and many configuration tool. But we are trying to design it better for container needs.

@hhorak the doc is lacking now, I'm working on update. the yaml file for instruction is for example:
``` yaml
- name: dummy
  description: execute dummy steps
  changes:
    - dummy:
      - dump: bar baz
      - dump: norfy
    - shell:
      - shell: pwd
      - shell: ls -l
      - shell: echo $HOME
```
which is just executing dummy module and its dump method with correct parameters. It also executes shell module shell method and run commands like pwd, ls, echo in that particular order. The main idea is that this yaml file will contain a configuration steps for particular state and is considered atomic.  So you can have yaml files like reconfigure jboss logging, setup database users. This change is just a set of modules to be executed and do the real stuff, like changing xml, invoking shell command, invoking jboss-cli and all of these fancy stuff. . 
This will enable you to share not only modules, but whole configuration recipes.

**Comment by l-d-j** [comment]
Also I've created https://gitter.im/containers-tools/cct for more general chat so we will not SPAM this issue
